internal/revision/scanner.go

```diff
@@ -0,0 +1,124 @@
+package revision
+
+import (
+	"bufio"
+	"io"
+	"unicode"
+)
+
+var zeroRune = rune(0)
+
+// scanner represents a lexical scanner.
+type scanner struct {
```

Comment:
The parser only use `newScanner()`, `scanner.scan()` and `scanner.unscan()` and the rest of the scanner functions (`scanner.read`, `scanner.unread`, `scanner.scanWord()`...) are helper functions used internally by scanner, but not by parser or any other type in the code.

This means you wrote `scanner` as to have an API composed of three functions:
- newScanner
- scan
- unscan
Which are the ones that are going to be used from the "outside" (in this case the parser). The rest of the functions are only going to be used from the inside and don't need so much error checks, for example.

But nobody knows that until the expend several minutes studying your code, which is wrong.

You can clarify those aspects by giving scanner an API, maybe even making it an internal package, so I don't have to spend time checking if a private method is doing extensive checking of its arguments, because clearly is just an internal thing.